Added a method to retrieve an Oauth2 Access Token using Google Play Services

bleathem opened the pull request:
This PR adds a method to the oauth2 cordova plugin enabling android applications to request an Oauth2 access token using one of the accounts authorized on the device.  The Google play Services API used is an Intent/Activity based one, and so we receive the access token via the cordova plugin's #onActivityResult method.

While this intent could have been initiated via an Aerogear authz module, the token still would have been returned via the #onActivityResult method, and the Authz module would have no knowledge of neither the token, nor the account used.  For these reasons the functionality was implemented solely within this plugin.

edewit commented:
:+1: I like it, what do you think @danielpassos

secondsun commented:
/me raises the Open Source Flag.
Two quest questions
1: The Manifest now has references to the gms packages which are not Open.  Will the affect the application if Google Play Services are not installed/linked at runtime/compile time?

2: Intent driven OAuth callbacks not not google play specific.  Would it be possible to rename the methods and refactor the code to be less reliant on Google Play and Google's nomenclature?

bleathem commented:
Thanks for the feedback @secondsun.  You're concerns seem well founded.

The code is meant to fail in cases where goolge play services is not available at runtime (for example on old devices).  The promised-based APi would then allow for the application to seek an Oauth2 token using traditional http means.

The current implementation however does place a hard-dependency at compile time.  We could re-structure the implementation to use a standard Intent interface, then use reflection to look for a particular implementation.  We could then have a configuration option to choose which implementation we look for via reflection.

If there are no objections I'll proceed with implementing this change.

danielpassos commented:
@edewit @bleathem @secondsun I will give it a try using the shot and share app this week

bleathem commented:
@danielpassos wrote:
> I will give it a try using the shot and share app this week

Does that mean you will try using this new feature and provide feedback? or that you will try implementing the suggested re-factor?

I'm keen on implementing the re-factor myself, I'd like to see this PR all the way through.  No sense in us duplicating the effort though, so please let me know if I should hold off.

secondsun commented:
@bleathem He means he will test the PR using a demo app.  Not he is going to implement it himself.

bleathem commented:
Awesome, thanks for the clarification @secondsun, sorry for the confusion @danielpassos !

bleathem commented:
Sadly the Google Play Services API does not solely use an Intent based API.  The token request process looks like:
1. `AccountManager#newChooseAccountIntent` - an Intent to select the account for which the token will be requested
2. `GoogleAuthUtil#getToken` - a Google play services API call to retrieve a token.  Note: this does not use an intent
3. The getToken call in 2) will throw a `UserRecoverableAuthException`, from which we request privileges (and consequently the token) using the `UserRecoverableException#getIntent` intent.

Even if we were to find a way to dynamically create the intents, we would still have the direct API call with which to contend.

The non-OSS approach of using Google Play Services still however seems like a blocker to me.  One solution may be to offer this functionality as a standalone non-supported cordova plugin.  Since this functionality currently does not integrate with the rest of the Aerogear OAuth approach, nothing would be lost.

bleathem commented:
Scratch my last comment.  In my head I conflated the ideas of using reflection and having the user specify which intent to use.  I've achieved the former, using reflection to invoke the gms class methods.  This allowed me to remove the gms dependency, however I still seem to need the _com.google.android.gms.version_ entry in the `AndroidManifest.xml` file to make it work.  It's not clear to me what the impact is of having this entry present.

As for de-coupling the API and/or having the user specify which intent to use, it's is unclear to me how I would proceed with that without a second intent-based oauth2 token request implementation.  So I will drop that idea for now, and leave the API google specific.

bleathem commented on plugin.xml:
```diff
@@ -50,7 +50,11 @@
       <service android:name="org.jboss.aerogear.android.authorization.oauth2.OAuth2AuthzService"/>
     </config-file>
+    <config-file target="AndroidManifest.xml" parent="/manifest/application">
+      <meta-data android:name="com.google.android.gms.version" android:value="@integer/google_play_services_version" />
```
Does anyone know the significance of this entry?  What effect will it have if google play services is not available?  Hopefully it will be safely ignored.

bleathem commented:
I removed the questionable gms addtion to the AndroidManifest.xml entry from the plugin.xml file.  There is no longer any direct reference to gms from this aerogear project.

Out-of-the box, an OAuth2 token request using google-play-services will fail with the callback message "Google Play Services is not available".  To enable Google Play Services one must add the `com.google.playservices` cordova plugin to one's project.  I updated the README with instructions indicating this.

edewit commented:
super stuff @bleathem I'll test it today

edewit commented:
After adding the `com.google.playservices` plugin I can no longer build the app with gradle:

```bash
/t/S/p/android ❯❯❯ gradle installDebug

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project 'android'.
> Configuration with name 'default' not found.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 5.512 secs
```

danielpassos commented:
:eyes:

danielpassos commented:
@edewit  Could you run that with --debug

I also had a problem but it was about Platform Tools version

```
* What went wrong:
Execution failed for task ':CordovaLib:preBuild'.
> failed to find Build Tools revision 19.0.0
```

I updated the `build.gradle` and `CordovaLib/build.gradle` to use the last Platform Tools version

```
compileSdkVersion 21
buildToolsVersion "21.1.1"
```

danielpassos commented:
Now I'm getting a problem with AndroidManifest.xml

```
* What went wrong:
Execution failed for task ':processDebugManifest'.
> Manifest merging failed. See console for more info.
```

```xml
<?xml version='1.0' encoding='utf-8'?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="org.jboss.aerogear.cordova" android:hardwareAccelerated="true" android:versionCode="1" android:versionName="0.0.1">
    <supports-screens android:anyDensity="true" android:largeScreens="true" android:normalScreens="true" android:resizeable="true" android:smallScreens="true" android:xlargeScreens="true" />
    <uses-permission android:name="android.permission.INTERNET" />
    <application android:hardwareAccelerated="true" android:icon="@drawable/icon" android:label="@string/app_name">
        <activity android:name="CordovaApp" android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale" android:label="@string/activity_name" android:launchMode="singleTop" android:theme="@android:style/Theme.Black.NoTitleBar" android:windowSoftInputMode="adjustResize">
            <intent-filter android:label="@string/launcher_name">
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
        <service android:name="org.jboss.aerogear.android.authorization.oauth2.OAuth2AuthzService" />
    </application>
    <uses-sdk android:minSdkVersion="10" android:targetSdkVersion="19" />
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
</manifest>
```

Not sure what is wrong.

bleathem commented:
Here's the steps I followed, with a project from scratch:

```bash
cordova create hello2 com.example.hello HelloWorld
cd hello2/
cordova platform add android
cordova plugin add ../aerogear-oauth2-cordova/
emacs platforms/android/build.gradle 
ANDROID_BUILD=gradle cordova run android
```

This yields the error:
```
/home/bleathem/workspace/cordova/hello2/platforms/android/cordova/node_modules/q/q.js:126
                    throw e;
                          ^
ERROR: Failed to launch application on device: ERROR: Failed to install apk to device: Error: Could not find apk architecture: arm build-type: debug
ERROR running one or more of the platforms: Error: /home/bleathem/workspace/cordova/hello2/platforms/android/cordova/run: Command failed with exit code 8
You may not have the required environment or OS to run this project
```

According to this [stackoverflow answer](http://stackoverflow.com/questions/28026717/cannot-run-cordova-application-on-device-or-emulators-error-running) this is caused by a bug in the cordova gradle build.  It's resolved in the current master branch.

```
cordova platform remove android
cordova platform add android@master --usegit
emacs platforms/android/build.gradle
ANDROID_BUILD=gradle cordova run android
```

The app then builds, use a chrome inspector to request an oauth2 token using google play services:
```
oauth2.requestAccessUsingPlayServices({scopes: 'openid'}).then(function(token) {console.log(token)}, function(error) {console.log(error)})
```

This will fail, because google play services is not available.  Add googleplay services:
```
cordova plugins add com.google.playservices
ANDROID_BUILD=gradle cordova run android
```

Again request an oauth2 plugin via google play services in chrome inspector. 

It works!

So depending on the latest master android platform is a problem.  Perhaps we can work around this if we remove our need to build with gradle. I have a few ideas of how we can accomplish that, perhaps I should bring that up on the aerogear mail list?

bleathem commented:
I've created a new PR (#16) that removes the gradle build requirement.  I've tested the result with this PR and it works well.

Further assessment of this PR should probably wait until #16 is merged.

bleathem commented:
I've closed PR #16 as it doesn't provide the necessary aar dependencies required for the webview dialog to work.

It seems then this PR is stuck with a dependency on the Cordova-Android master branch.  How do we want to proceed?  Shall we leave this PR hanging, and revisit once Cordova-Android 4.0 has been released?

edewit commented:
I have one more idea to get rid of the gradle dependency let me give that a try and otherwise we'll just keep this open until 4.0 lands, can't be long as they are already taking votes on the matter.

edewit commented:
with #17 merged we can fix merge this one as well right @bleathem ?

bleathem commented:
Yes.  I'll rebase, test, and update this PR.

bleathem commented:
I've rebased my changes against the master branch, and tested that the Oauth2 token request works as expected.

IMO the following blockers for this PR have been overcome:
1) The compile-time and run-time dependencies on the google gms API is now optional
2) The Cordova-Android@master platfrom requirement is not longer necessary

The API introduced continues to use a naming convention tightly coupled to google, but this reflects that the implementation (while optional) is itself tightly coupled to google.

Updated steps to test this new functionality are:

```bash
cordova create hello2 com.example.hello HelloWorld
cd hello2/
cordova platform add android
cordova plugin add ../aerogear-oauth2-cordova/
cordova run android
```

The app then builds, use a chrome inspector to request an oauth2 token using google play services:
```
oauth2.requestAccessUsingPlayServices({scopes: 'openid'}).then(function(token) {console.log(token)}, function(error) {console.log(error)})
```

This will fail as expected, with the error message: "Google Play Services is not available".  Add googleplay services:
```
cordova plugins add com.google.playservices
cordova run android
```

Again request an oauth2 plugin via google play services in chrome inspector. 

It works!

edewit commented:
Tested and works like a charm

edewit commented:
thanks @bleathem for you perseverance and contribution
